crates/red_knot_python_semantic/src/semantic_index/predicate.rs

```diff
@@ -85,3 +87,60 @@ impl<'db> PatternPredicate<'db> {
         self.file_scope(db).to_scope_id(db, self.file(db))
     }
 }
+
+/// A "placeholder predicate" that is used to model the fact that the boundness of a
+/// (possible) definition or declaration caused by a `*` import cannot be fully determined
+/// until type-inference time. This is essentially the same as a standard visibility constraint,
+/// so we reuse the [`Predicate`] infrastructure to model it.
+///
+/// To illustrate, say we have a module `a.py` like so:
+///
+/// ```py
+/// if <condition>:
+///     class A: ...
+/// ```
+///
+/// and we have a module `b.py` like so:
+///
+/// ```py
+/// A = 1
+///
+/// from a import *
+/// ```
+///
+/// Since we cannot know whether or not <test> is true at semantic-index time,
+/// we record a definition for `A` in `b.py` as a result of the `from a import *`
+/// statement, but place a predicate on it to record the fact that we don't yet
+/// know whether this definition will be visible from all control-flow paths or not.
+/// Essentially, we model `b.py` as something similar to this:
+///
+/// ```py
+/// A = 1
+///
+/// if <star_import_placeholder_predicate>:
+///     from a import A
+/// ```
+///
+/// At type-check time, the placeholder predicate for the `A` definition is evaluated by
+/// attempting to resolve the `A` symbol in `a.py`'s global namespace:
+/// - If it resolves to a definitely bound symbol, then the predicate resolves to [`Truthiness::AlwaysTrue`]
+/// - If it resolves to an unbound symbol, then the predicate resolves to [`Truthiness::AlwaysFalse`]
+/// - If it resolves to a possibly bound symbol, then the predicate resolves to [`Truthiness::Ambiguous`]
+///
+/// [Truthiness]: [crate::types::Truthiness]
+#[salsa::tracked(debug)]
+pub(crate) struct StarImportPlaceholderPredicate<'db> {
+    pub(crate) importing_file: File,
+
+    pub(crate) symbol_id: ScopedSymbolId,
```

Comment:
This is a scoped symbol ID, meaning it is only valid in some specific scope, and can give "random" wrong results if interpreted as a symbol ID in some other scope.

It's made clear in a comment below that the scope is implicitly always "global scope of importing_file", but I think it would be worth mentioning that in a doc comment here as well.